File: leetcode/1001-2000/1401-1500/1441-1450/1448-count-good-nodes-in-binary-tree/CountGoodNodesInBinaryTree.java
public class CountGoodNodesInBinaryTree {
  public static void main(String[] args) {
    TreeNode root = new TreeNode(3, new TreeNode(3, new TreeNode(4), new TreeNode(2)), null);

    System.out.println(new Solution().goodNodes(root));
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  int goodNodes(TreeNode node, int max) {
    if (node == null) {
      return 0;
    }

    return (node.val >= max ? 1 : 0) + goodNodes(node.left, Math.max(max, node.val))
        + goodNodes(node.right, Math.max(max, node.val));
  }

  public int goodNodes(TreeNode root) {
    return 1 + goodNodes(root.left, root.val) + goodNodes(root.right, root.val);
  }
}

File: leetcode/1001-2000/1401-1500/1441-1450/1448-count-good-nodes-in-binary-tree/CountGoodNodesInBinaryTree.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function goodNodes(root: TreeNode | null): number {
  if (root == null) {
    return 0;
  }

  const queue: { node: TreeNode, max: number }[] = [{ node: root, max: Number.NEGATIVE_INFINITY }];
  let totalGoodNodes = 0;

  while (queue.length > 0) {
    let { node, max } = queue.shift()!;

    if (node.val >= max) {
      totalGoodNodes++;
    }

    max = Math.max(max, node.val);

    if (node.right != null) {
      queue.unshift({ node: node.right, max });
    }

    if (node.left != null) {
      queue.unshift({ node: node.left, max })
    }
  }

  return totalGoodNodes;
};

console.log(goodNodes(new TreeNode(3, new TreeNode(3, new TreeNode(4), new TreeNode(2)), null)));
